File: T-Bank/Algorithms-and-Data-Structures-2025/11sprint-Graph2/contest/B-Остовное_дерево/solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <tuple>
#include <algorithm>


class UnionFind {
private:
    int n;
    std::vector<int> parent;
    std::vector<int> rank;
    std::vector<int> minValue;
    std::vector<int> maxValue;

public:
    long long totalWeight = 0;
    UnionFind(int n_) : n(n_) {
        parent.resize(n, 0);
        rank.resize(n, 1);
        minValue.resize(n, 0);
        maxValue.resize(n, 0);
        for (int i = 0; i < n; ++i) {
            parent[i] = i;
            minValue[i] = i;
            maxValue[i] = i;
        }
    }
    
    int find(int x) {
        while (x != parent[x]) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }
    
    void union_(int v, int u, int weight) {
        int vParent = find(v), uParent = find(u);
        if (vParent == uParent) return;
        totalWeight += weight;
        if (rank[vParent] >= rank[uParent]) {
            parent[uParent] = vParent;
            rank[vParent] += rank[uParent];
            minValue[vParent] = std::min(minValue[vParent], minValue[uParent]);
            maxValue[vParent] = std::max(maxValue[vParent], maxValue[uParent]);
        } else {
            parent[vParent] = uParent;
            rank[uParent] += rank[vParent];
            minValue[uParent] = std::min(minValue[uParent], minValue[vParent]);
            maxValue[uParent] = std::max(maxValue[uParent], maxValue[vParent]);
        }
    }
    
    void get(int x) {
        int xParent = find(x);
        std::printf("%d %d %d\n", minValue[xParent], maxValue[xParent], rank[xParent]);
    }
};



void solution() {
    /*
    Time Complexity O(MlogN)
    Memory Complexity O(N)
    */
    int n, m;
    std::scanf("%d %d", &n, &m);
    UnionFind uf(n + 1);
    std::vector<std::tuple<int, int, int>> edges;
    for (int i = 0; i < m; ++i) {
        int source, destination, weight;
        std::scanf("%d %d %d", &source, &destination, &weight);
        edges.push_back({weight, source, destination});
    }
    std::sort(edges.begin(), edges.end());
    for (std::tuple<int, int, int>& edge : edges) {
        uf.union_(std::get<1>(edge), std::get<2>(edge), std::get<0>(edge));
    }
    std::printf("%lld\n", uf.totalWeight);
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/11sprint-Graph2/contest/B-Остовное_дерево/solution.py
import sys


class UnionFind:
    def __init__(self, n: int) -> None:
        self.parent: list[int] = list(range(n))
        self.rank: list[int] = [1] * n
        self.min_value: list[int] = list(range(n))
        self.max_value: list[int] = list(range(n))
        self.total_weight: int = 0
    
    def find(self, x: int) -> int:
        while x != self.parent[x]:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x
    
    def union_(self, u: int, v: int, weight: int) -> None:
        u_parent: int = self.find(u)
        v_parent: int = self.find(v)
        if u_parent == v_parent: return
        self.total_weight += weight
        if self.rank[u_parent] >= self.rank[v_parent]:
            self.parent[v_parent] = u_parent
            self.rank[u_parent] += self.rank[v_parent]
            self.min_value[u_parent] = min(self.min_value[u_parent], self.min_value[v_parent])
            self.max_value[u_parent] = max(self.max_value[u_parent], self.max_value[v_parent])
        else:
            self.parent[u_parent] = v_parent
            self.rank[v_parent] += self.rank[u_parent]
            self.min_value[v_parent] = min(self.min_value[v_parent], self.min_value[u_parent])
            self.max_value[v_parent] = max(self.max_value[v_parent], self.max_value[u_parent])
    
    def get(self, x: int) -> None:
        x_parent: int = self.find(x)
        sys.stdout.write('{} {} {}\n'.format(self.min_value[x_parent], self.max_value[x_parent], self.rank[x_parent]))


def solution() -> None:
    """
    Time Complexity O(MlogN)
    Memory Complexity O(N)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    uf: UnionFind = UnionFind(n + 1)
    edges: list[tuple[int, int, int]] = []
    for _ in range(m):
        edge: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
        edges.append((edge[2], edge[0], edge[1]))
    edges.sort()
    for edge in edges:
        uf.union_(edge[1], edge[2], edge[0])
    sys.stdout.write('{}\n'.format(uf.total_weight))


if __name__ == '__main__':
    solution()